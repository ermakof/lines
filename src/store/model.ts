import IUserProfile from '@src/modules/Auth/model/IUserProfile';

export interface IState {
  isLoading?: boolean;
  gameFieldSize: number;
  gameFieldPercentFilled: number;
  gameFieldData: Array<number>;
  selectedCell?: number;
  userLevel: string;
  userProfile?: IUserProfile;
}
