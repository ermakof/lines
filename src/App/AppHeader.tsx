import React, { FC, memo } from 'react';
import Panel from '@src/layout/Panel';
import AppRouter from '@src/App/AppRouter';
import styled from '@emotion/styled';

const Root = styled(Panel)`
  width: 100%;
`;

const AppHeader: FC = () => (
  <Root role="topPanel">
    <AppRouter />
  </Root>
);

export default memo(AppHeader);
