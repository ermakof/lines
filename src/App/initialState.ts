import { IAppState } from '@src/App/model/IAppState';
import createGameField from '@src/utils/createGameField';

export const GAME_FIELD_SIZE = 9;

export const INIT_GAME_LEVEL = '1';
export const INIT_GAME_FIELD_PERCENT_FILLED = 0;

interface ISettings {
  percent: number;
  chain: number;
  increase: number;
}
interface IProps {
  [key: string]: ISettings;
}
export const GAME_LEVEL_SETTINGS: IProps = {
  '1': { percent: 10, chain: 2, increase: 3 },
  '2': { percent: 20, chain: 3, increase: 4 },
  '3': { percent: 30, chain: 4, increase: 5 },
};

const initialState: IAppState = {
  userLevel: INIT_GAME_LEVEL,
  gameFieldPercentFilled: 0,
  gameFieldData: createGameField(GAME_FIELD_SIZE ** 2, INIT_GAME_FIELD_PERCENT_FILLED),
  highlightedCells: undefined,
  score: 0,
  selectedCell: undefined,
  isLoading: false,
  hitParade: undefined,
};

export default initialState;
