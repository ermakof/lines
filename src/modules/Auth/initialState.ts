import { IAuthState } from '@src/modules/Auth/model/IAuthState';

const initialState: IAuthState = {
  userProfile: undefined,
};

export default initialState;
