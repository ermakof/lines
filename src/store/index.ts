import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';

import rootReducer from '@src/store/rootReducer';

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware()
      .prepend
      // correctly typed middlewares can just be used
      ()
      // prepend and concat calls can be chained
      .concat(logger),
});

type TRootState = ReturnType<typeof store.getState>;

export { store, TRootState };
