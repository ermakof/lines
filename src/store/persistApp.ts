import { LOCAL_STORAGE_APP_KEY, store } from '@src/store/index';

const persistApp = () => {
  const app = store.getState().app;
  localStorage.setItem(LOCAL_STORAGE_APP_KEY, JSON.stringify(app));
};

export default persistApp;
