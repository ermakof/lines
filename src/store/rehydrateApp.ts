import { actions } from '@src/modules/Auth/authSlice';
import { store } from '@src/store/index';

export const rehydrateApp = () => {
  store.dispatch(actions.rehydrate());
};
