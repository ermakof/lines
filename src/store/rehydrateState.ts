import appSlice from '@src/App/appSlice';
import { LOCAL_STORAGE_APP_KEY, store } from '@src/store/index';

const rehydrateState = () => {
  try {
    const persistedState = localStorage.getItem(LOCAL_STORAGE_APP_KEY);
    if (persistedState) {
      const payload = JSON.parse(persistedState);
      store.dispatch(appSlice.actions.hydrate(payload));
    }
  } catch (e) {
    console.log(e);
  }
};

export default rehydrateState;
