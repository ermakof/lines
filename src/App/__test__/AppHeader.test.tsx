import React from 'react';
import { Provider, useDispatch, useSelector } from 'react-redux';
import { render, screen } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';

import AppHeader from '@src/App/AppHeader';
import initialState from '@src/App/initialState';
import { TRootState } from '@src/store';

const mockStore = configureMockStore();

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

let store: TRootState;
let mockDispatch;

describe('AppHeader', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    mockDispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  it('Render authorised <AppHeader>', () => {
    (useSelector as jest.Mock).mockReturnValue({
      userProfile: { login: '123', password: '123', token: '7a0cbf93-aa13-4b50-8a41-97ddbfba00d5' },
    });
    const { asFragment } = render(
      <Provider store={store}>
        <AppHeader />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const title = screen.getByText(/Игрок/gi);
    expect(title).toBeInTheDocument();
    const button = screen.getByRole(/^buttonLogout$/gi);
    expect(button).toBeInTheDocument();
    const select = screen.getByRole(/select/gi);
    expect(select).toBeInTheDocument();
  });

  it('Render unauthorised <AppHeader>', () => {
    (useSelector as jest.Mock).mockReturnValue({});
    const { asFragment } = render(
      <Provider store={store}>
        <AppHeader />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const title = screen.getByText(/Lines/gi);
    expect(title).toBeInTheDocument();
    const button = screen.getByRole(/buttonLogin/gi);
    expect(button).toBeInTheDocument();
  });
});
