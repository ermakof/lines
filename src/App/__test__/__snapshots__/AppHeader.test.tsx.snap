// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader Render authorised <AppHeader> 1`] = `
<DocumentFragment>
  <div
    class="css-1w0hfqs"
    role="topPanel"
  >
    <div
      class="css-ov1ktg"
    >
      <p
        class="css-10vuvr5"
      >
        Игрок: 123
      </p>
      <div
        class="css-1w0hfqs"
        role="userForm"
      >
        <div
          class="css-ov1ktg"
        >
          <button
            class="css-tv5zn1"
            role="buttonReset"
          >
            Сброс
          </button>
          <div
            class="css-1xvr7bb"
          >
            <p>
              Уровень
            </p>
            <select
              name="select"
              role="select"
            >
              <option
                role="option"
                value="1"
              >
                Падаван
              </option>
              <option
                role="option"
                value="2"
              >
                Джедай
              </option>
              <option
                role="option"
                value="3"
              >
                Член Совета Силы
              </option>
            </select>
          </div>
          <button
            class="css-tv5zn1"
            role="buttonLogout"
          >
            Выйти
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppHeader Render unauthorised <AppHeader> 1`] = `
<DocumentFragment>
  <div
    class="css-1w0hfqs"
    role="topPanel"
  >
    <div
      class="css-ov1ktg"
    >
      <p
        class="css-10vuvr5"
      >
        Lines
      </p>
      <div
        class="css-9e74ss"
        role="authForm"
      >
        <form
          action="#"
        >
          <label
            for="login"
          >
            Имя
          </label>
          <input
            id="login"
            name="login"
            placeholder="Введите логин"
            role="login"
            value=""
          />
          <label
            for="password"
          >
            Пароль
          </label>
          <input
            id="password"
            name="password"
            placeholder="Введите пароль"
            role="password"
            value=""
          />
          <button
            disabled=""
            role="buttonLogin"
            type="submit"
          >
            Войти
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
